Feature/add mesh guide (#4081)
fishtown-analytics/docs.getdbt.com: website/docs/guides/best-practices/how-we-mesh/mesh-3-implementation.md

---
title: "Implementing your mesh plan"
description: Getting started with dbt Mesh patterns
hoverSnippet: Learn how to get started with dbt Mesh
---

> As mentioned before, the key decision to migrating to a multi-project architecture is understanding how your project is already being grouped, built, and deployed. We can use this information to inform our decision to split our project apart.

Reviewer: suggest centering on second person
```suggestion
As mentioned before, the key decision to migrating to a multi-project architecture is understanding how your project is already being grouped, built, and deployed. You can use this information to inform your decision on how to split your project apart.
```

- **Examine your jobs** which sets of models are most often built together?
- **Look at your lineage graph** how are models connected?

> - **Look at your selectors** defined in `selectors.yml` - how are people already defining resource groups?

Reviewer: ```suggestion
- **Look at your selectors** defined in `selectors.yml` - how do people already define resource groups?
```

> - **Talk to teams** about what sort of separation is naturally existing right now.

Reviewer: ```suggestion
- **Talk to teams** about what sort of separation naturally exists right now.
```

  - Are there various domains people are focused on?
  - Are there various sizes, shapes, and sources of data that get handled separately (such as click event data)?
  - Are there people focused on separate levels of transformation, such as landing and staging data or building marts?

## Add groups and access

> Once you have a sense of some initial groupings, the first step is to implement **group and access permissions** within a single project.

Reviewer: ```suggestion
Once you have a sense of some initial groupings, you can first implement **group and access permissions** within a single project.
```

> - First we'll create a [group](/docs/build/groups) to define the owner of a set of models.

Reviewer: ```suggestion
- First you can create a [group](/docs/build/groups) to define the owner of a set of models.
```

```yml
# in models/__groups.yml

groups: 
  - name: marketing
    owner:
      - name: Ben Jaffleck 
        email: ben.jaffleck@jaffleshop.com
```

> - Then, we can add models to that group using the `group:` key in the model's yml entry.

Reviewer: ```suggestion
- Then, add models to that group using the `group:` key in the model's yml entry.
```

```yml
# in models/marketing/__models.yml

models: 
  - name: fct_marketing_model
    group: marketing
  - name: stg_marketing_model
    group: marketing
```

> - Once models are added to the group, we will **add [access](/docs/collaborate/govern/model-access) settings to the models** based on their connections between groups, *opting for the most private access that will maintain current functionality*. This means that any model that has *only* relationships to other models in the same group should be `private` , and any model that has cross-group relationships, or is a terminal node in the group DAG should be `protected` so that other parts of the DAG can continue to reference it.

Reviewer: ```suggestion
- Once you've added models to the group, you can **add [access](/docs/collaborate/govern/model-access) settings to the models** based on their connections between groups, *opting for the most private access that will maintain current functionality*. This means that any model that has *only* relationships to other models in the same group should be `private` , and any model that has cross-group relationships, or is a terminal node in the group DAG should be `protected` so that other parts of the DAG can continue to reference it.
```

1. If using `source` method:
   1. In the parent project:
      1. mark all models being imported downstream as `public` and add a model contract.
   2. In the child project:
      1. Add the upstream project to your `dependencies.yml`
      2. Replace the `{{ source() }}` functions with cross project `{{ ref() }}` functions.
      3. Remove the unnecessary `source` definitions.

## Additional Resources
### Our example projects

We've provided a set of example projects you can use to explore the topics covered here. We've split our [Jaffle Shop](https://github.com/dbt-labs/jaffle-shop) project into 3 separate projects in a multi-repo dbt Mesh. Note that you'll need to leverage dbt Cloud to use multi-project architecture, as cross-project references are powered via dbt Cloud's APIs.

- **[Platform](https://github.com/dbt-labs/jaffle-shop-mesh-platform)** - containing our centralized staging models.
- **[Marketing](https://github.com/dbt-labs/jaffle-shop-mesh-marketing)** - containing our marketing marts.
- **[Finance](https://github.com/dbt-labs/jaffle-shop-mesh-finance)** - containing our finance marts.

### dbt-meshify

> We recommend using the `dbt-meshify` [command line tool](<https://dbt-labs.github.io/dbt-meshify/0.1/>) to help you do this. This comes with CLI operations to automate most of the above steps.

Reviewer: by CLI, do we mean dbt Cloud CLI or the command line interface in general?

Reviewer 2: command line interface in general - dbt-meshify is separate from the Cloud CLI

Reviewer: We've been moving away from CLI except when using Cloud CLI, so we might want to do the same here

---
title: "Implementing your mesh plan"
description: Getting started with dbt Mesh patterns
hoverSnippet: Learn how to get started with dbt Mesh
---

## Implementing a dbt Mesh

Let's examine an outline of steps to start implementing a dbt Mesh in your organization.

### Research your current structure

- Look at your selectors to figure out how people are grouping models right now.
- Talk to teams about what sort of separation is naturally existing right now
  - Are there various domains people are focused on?

>   - Are there various sizes, shapes, and sources of data that get handled separately (such as click event data)?

Reviewer 3: handled could be expanded to mean developed + deployed + managed ?

  - Are there people focused on separate levels of transformation, such as landing and staging data or building marts?

## Add groups and access

Once you have a sense of some initial groupings, implement group and access permissions within a project.

> - Incrementally start building your jobs based on these groups (we would recommend in parallel to your production jobs until you’re sure about them) to feel out that you’ve drawn the lines in the right place.

Reviewer 3: might be worth calling out this can/should be done in a QA deployment environment

## Do the splits

> - When you’ve confirmed the right groups, use `dbt-meshify` to pull chunks out into their own projects.

Reviewer 3: ```suggestion
- When you’ve confirmed the right groups, use `dbt-meshify` ([source](https://github.com/dbt-labs/dbt-meshify)) to pull chunks out into their own projects.
```
